### DISCUSSION TITLE
Undeprecate git.io and restrict new URLs to `github.com` domain

### DISCUSSION THREAD

**Discussion by Alhadis** [community/community#11796] [Other Feature Feedback, Questions, & Ideas] [answered]
Last January, GitHub announced[^1^] on their blog that they'd be retiring their URL-shortening service, `git.io`, citing reasons of malicious/inappropriate use:

> Today, git.io is increasingly being used for malicious purposes. At GitHub, we want to end that activity, focus on building great developer tools, and cede URL shortening to companies and teams who provide it as a core offering.
> […]
> For these reasons, we have disallowed new link creation on git.io. Existing URLs will continue to be accessible, but we encourage using one of the many URL shortening services that are available, instead of git.io, as we will be deprecating the tool in the future.

Instead of retiring the service completely, why not restrict all new shortlinks to `github.com`-hosted URLs only? I can understand not wanting users to mask links to `github.io` URLs (since a malicious binary can always be uploaded to <code><var>username</var>.github.io/NOT_A_VIRUS.LOL.EXE</code> or whatever).

Heck, I was surprised to learn that was ever an option: this whole time I had the belief that `git.io` generated shortlink IDs based on resource, not URL contents; i.e., the following two URLs would both produce the same shortlink ID:

```
https://github.com/mozilla/gecko-dev/blob/7e953e97a3a/testing/marionette/client/marionette_driver/geckoinstance.py
https://github.com/mozilla/gecko-dev/blob/7e953e97a3ab77e39cdbf3140164e7843f3f026a/testing/marionette/client/marionette_driver/geckoinstance.py
```

Anyway, it's a shame to lose such a handy service; `git.io` has made many a lengthy permalink fit the bodies of a 72-column commit message. 😉

[^1^]: https://github.blog/changelog/2022-01-11-git-io-no-longer-accepts-new-urls/

**Comment by dhimmel** [comment]
Just came across [this post](https://github.blog/changelog/2022-04-25-git-io-deprecation/) that:

> Effective Friday, April 29, 2022 all links on git.io will stop redirecting. Please update any existing links that make use of the git.io URL service immediately.

I've been using git.io links for almost a decade. Now I have go through hundreds of repos and external content and scramble to update links in just a few days time.

Seems like a terrible amount of link rot is headed our way. I've used git.io links in many academic works where long term preservation is essential. Seems crazy that old links will stop redirecting!

GitHub should have scanned repos for these links and created PRs to replace them. Most users who have git.io links are probably unaware of the impending disaster.

**Comment by Alhadis** [comment]
> Seems like a terrible amount of link rot is headed our way. I've used git.io links in many academic works where long term preservation is essential.

Likewise. I (wrongly) assumed that for as long as GitHub existed as a service, so too would `git.io`. Now I not only need to update shortened URLs in each of my projects, I also need to run them through `web.archive.org` to ensure that those I can't change (e.g., those in commit messages) are at least semi-recoverable. Neither of which are things I'd rather be doing.

Marking this as an answer, since it more-or-less confirms that my original request is impossible.

**Reply by dhimmel** [reply]
> I (wrongly) assumed that for as long as GitHub existed as a service, so too would git.io.

Exactly. I assumed git.io links were as good as GitHub!

**Reply by Alhadis** [reply]
[Saved by the bell!](https://github.blog/changelog/2022-04-25-git-io-deprecation/) Looks like we're not alone:

> **2022-04-27 Update:**
We have received feedback from developers and academic researchers who have published git.io links in print documentation and research papers. In order to preserve the integrity of these historical documents, we have decided to archive the current git.io links in a new read-only service that will allow us to serve redirects for those links longer term.

**Reply by dhimmel** [reply]
Ah amazing! This updated policy seems a lot more sensible... remove any malicious links while continuing to redirect existing ones.

**Comment by ThiefMaster** [comment]
> Anyway, it's a shame to lose such a handy service; git.io has made many a lengthy permalink fit the bodies of a 72-column commit message. 😉

At least for this part, you can use `user/repo@commithash` - not as short but more user-friendly and not excessively long either.

**Comment by queenofcorgis** [comment, marked as answer]
**2022-04-27 Update:**
While the git.io url redirection service is [read-only](https://github.blog/2011-11-10-git-io-github-url-shortener/) and use of the service is limited, we have received feedback from developers and academic researchers who have published git.io links in print documentation and research papers. In order to preserve the integrity of these historical documents, we have decided to archive the current git.io links in a new read-only service that will allow us to serve redirects for those links longer term.

As we continue our analysis, we may remove individual links that point to spammy, malicious or 404 links. Our goal is to not break links relied on for legitimate use, especially by the academic community, while preserving the security of developers on GitHub.

That said, we still encourage users to make use of one of the many URL shortening services available with greater functionality than the git.io service provided. GitHub support will not be able to update or edit redirection records served by the git.io archive service. [View the post here](https://github.blog/changelog/2022-04-25-git-io-deprecation/).
